File: sdk/node/bes.js
// bes Node.js SDK — JSON-RPC over HTTP client
// No gRPC dependency needed.
//
// 用法一（推荐）：静态工厂，返回已就绪实例
//   const sandbox = await Sandbox.create({ serverAddr: 'localhost:19821' });
//   await sandbox.eval('navigator.userAgent');
//
// 用法二：构造后 await ready
//   const sandbox = new Sandbox();
//   await sandbox.ready;
//   await sandbox.eval('navigator.userAgent');

class Sandbox {
  // Async factory: resolves once the session is created and usable.
  static async create(opts = {}) {
    const { serverAddr = 'localhost:19821', ...rest } = opts;
    const s = new Sandbox(serverAddr, rest);
    await s._ready;
    return s;
  }

  constructor(serverAddr = 'localhost:19821', opts = {}) {
    this.baseURL = `http://${serverAddr}`;
    this.sessionId = null;
    // D5 fix: the constructor fires session creation async; every API awaits
    // this promise first so `new Sandbox(...).eval(...)` no longer races
    // "Session not created".
    this._ready = this._createSession(opts);
  }

  async _post(path, data) {
    const resp = await fetch(`${this.baseURL}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(data),
    });
    if (!resp.ok) {
      const text = await resp.text();
      throw new Error(`HTTP ${resp.status}: ${text}`);
    }
    return resp.json();
  }

  async _get(path) {
    const resp = await fetch(`${this.baseURL}${path}`);
    if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
    return resp.json();
  }

  async _delete(path) {
    const resp = await fetch(`${this.baseURL}${path}`, { method: 'DELETE' });
    return resp.json();
  }

  async _createSession(opts) {
    const { browser = 'chrome', os = 'windows', seed = 0, location = 'https://example.com/',
            cookies = {}, proxy = '', netMode = 'live', recording = '' } = opts;
    const resp = await this._post('/api/session', {
      browser, os, seed, location, cookies, proxy, net_mode: netMode, recording,
    });
    if (resp.error) throw new Error(resp.error);
    this.sessionId = resp.session_id;
    return this;
  }

  async _ensureSession() {
    await this._ready;
    if (!this.sessionId) throw new Error('Session not created');
  }

  async eval(code) {
    await this._ensureSession();
    const resp = await this._post(`/api/session/${this.sessionId}/eval`, { code });
    if (resp.error) throw new Error(resp.error);
    return resp.result || '';
  }

  async loadScript(name, content) {
    await this._ensureSession();
    if (content === undefined) {
      const fs = require('fs');
      content = fs.readFileSync(name, 'utf-8');
    }
    const resp = await this._post(`/api/session/${this.sessionId}/script`, { name, content });
    if (resp.error) throw new Error(resp.error);
  }

  async call(functionName, ...args) {
    await this._ensureSession();
    const resp = await this._post(`/api/session/${this.sessionId}/call`, {
      function_name: functionName, args,
    });
    if (resp.error) throw new Error(resp.error);
    return resp.result || '';
  }

  async getFingerprint() {
    await this._ensureSession();
    return this._get(`/api/session/${this.sessionId}/fingerprint`);
  }

  async getCookies() {
    await this._ensureSession();
    const resp = await this._get(`/api/session/${this.sessionId}/cookies`);
    return resp.cookies || '';
  }

  async setCookie(name, value) {
    await this._ensureSession();
    await this._post(`/api/session/${this.sessionId}/cookies`, { name, value });
  }

  async close() {
    await this._ready.catch(() => {});
    if (this.sessionId) {
      try { await this._delete(`/api/session/${this.sessionId}`); } catch (e) {}
      this.sessionId = null;
    }
  }
}

module.exports = { Sandbox };

File: sdk/go/bes.go
// Package bes provides a Go SDK for browser-env-sandbox.
// Connects to bes-server via JSON-RPC over HTTP.
package bes

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// Sandbox wraps a connection to a bes-server.
type Sandbox struct {
	baseURL   string
	client    *http.Client
	sessionID string
}

// New creates a new Sandbox client.
func New(serverAddr string) *Sandbox {
	return &Sandbox{
		baseURL: "http://" + serverAddr,
		client: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
}

// SessionOptions configures a session.
type SessionOptions struct {
	Browser  string            `json:"browser"`
	OS       string            `json:"os"`
	Seed     uint64            `json:"seed"`
	Location string            `json:"location"`
	Cookies  map[string]string `json:"cookies"`
	Proxy    string            `json:"proxy"`
	NetMode  string            `json:"net_mode"`
}

// CreateSession creates a sandbox session.
func (s *Sandbox) CreateSession(opts SessionOptions) error {
	resp, err := s.post("/api/session", opts)
	if err != nil {
		return err
	}
	s.sessionID = resp["session_id"].(string)
	return nil
}

// Eval executes JavaScript and returns the result.
func (s *Sandbox) Eval(code string) (string, error) {
	resp, err := s.post(fmt.Sprintf("/api/session/%s/eval", s.sessionID), map[string]string{"code": code})
	if err != nil {
		return "", err
	}
	if e, ok := resp["error"].(string); ok && e != "" {
		return "", fmt.Errorf("%s", e)
	}
	result, _ := resp["result"].(string)
	return result, nil
}

// LoadScript loads and executes a script.
func (s *Sandbox) LoadScript(name, content string) error {
	_, err := s.post(fmt.Sprintf("/api/session/%s/script", s.sessionID), map[string]string{
		"name":    name,
		"content": content,
	})
	return err
}

// Call calls a global function by name.
func (s *Sandbox) Call(functionName string, args ...string) (string, error) {
	resp, err := s.post(fmt.Sprintf("/api/session/%s/call", s.sessionID), map[string]interface{}{
		"function_name": functionName,
		"args":          args,
	})
	if err != nil {
		return "", err
	}
	if e, ok := resp["error"].(string); ok && e != "" {
		return "", fmt.Errorf("%s", e)
	}
	result, _ := resp["result"].(string)
	return result, nil
}

// GetFingerprint returns the session fingerprint.
func (s *Sandbox) GetFingerprint() (map[string]interface{}, error) {
	return s.get(fmt.Sprintf("/api/session/%s/fingerprint", s.sessionID))
}

// GetCookies returns the session cookies.
func (s *Sandbox) GetCookies() (string, error) {
	resp, err := s.get(fmt.Sprintf("/api/session/%s/cookies", s.sessionID))
	if err != nil {
		return "", err
	}
	c, _ := resp["cookies"].(string)
	return c, nil
}

// SetCookie sets a cookie.
func (s *Sandbox) SetCookie(name, value string) error {
	_, err := s.post(fmt.Sprintf("/api/session/%s/cookies", s.sessionID), map[string]string{"name": name, "value": value})
	return err
}

// Close closes the session.
func (s *Sandbox) Close() error {
	if s.sessionID == "" {
		return nil
	}
	req, err := http.NewRequest("DELETE", fmt.Sprintf("%s/api/session/%s", s.baseURL, s.sessionID), nil)
	if err != nil {
		return err
	}
	_, err = s.client.Do(req)
	s.sessionID = ""
	return err
}

// --- HTTP helpers ---

func (s *Sandbox) post(path string, body interface{}) (map[string]interface{}, error) {
	data, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest("POST", s.baseURL+path, bytes.NewReader(data))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	return s.do(req)
}

func (s *Sandbox) get(path string) (map[string]interface{}, error) {
	req, err := http.NewRequest("GET", s.baseURL+path, nil)
	if err != nil {
		return nil, err
	}
	return s.do(req)
}

func (s *Sandbox) do(req *http.Request) (map[string]interface{}, error) {
	resp, err := s.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("parse response failed: %w (body: %s)", err, string(data))
	}
	return result, nil
}
